File: keychain/src/networks/ethereum/key_path.rs
use key_path::{Error, KeyPath as IKeyPath, BIP44_PURPOSE, BIP44_SOFT_UPPER_BOUND};

/// the BIP44 coin type for Ethereum
pub const BIP44_COIN_TYPE: u32 = 0x8000003c;

#[derive(Debug, Copy, Clone)]
pub struct KeyPath {
  account: u32,
  address: u32
}

impl KeyPath {
  pub fn new(account: u32) -> Result<Self, Error> {
    if account >= BIP44_SOFT_UPPER_BOUND {
      return Err(Error::InvalidAccount(account));
    }
    Ok(KeyPath { account: account + BIP44_SOFT_UPPER_BOUND, address: 0 })
  }

  pub fn new_metamask(account: u32) -> Result<Self, Error> {
    if account >= BIP44_SOFT_UPPER_BOUND {
      return Err(Error::InvalidAccount(account));
    }
    Ok(KeyPath { account: BIP44_SOFT_UPPER_BOUND, address: account })
  }
}

impl IKeyPath for KeyPath {
  fn purpose(&self) -> u32 {
    BIP44_PURPOSE
  }

  fn coin(&self) -> u32 {
    BIP44_COIN_TYPE
  }

  fn account(&self) -> u32 {
    self.account
  }

  fn change(&self) -> u32 {
    0
  }

  fn address(&self) -> u32 {
    self.address
  }
}

File: keychain-java/src/main/java/one/tesseract/keychain/ethereum/KeyPath.java
package one.tesseract.keychain.ethereum;

import one.tesseract.keychain.IKeyPath;
import one.tesseract.keychain.RustObject;

public class KeyPath extends RustObject implements IKeyPath{
  public KeyPath(long ptr) {
    super(ptr);
  }

  public static native KeyPath newKeyPath(int account);
  public static native KeyPath newMetamask(int account);

  @Override
  public native int purpose();

  @Override
  public native int coin();

  @Override
  public native int account();

  @Override
  public native int change();

  @Override
  public native int address();

  @Override
  public native void free();
}
